applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,744 @@
+			chmap_inst,
+			evt->channel_index,
+			evt->crc_ok_count,
+			evt->crc_error_count);
+		return true;
+	default:
+		return false;
+	}
+}
+
+static void enable_qos_reporting(void)
+{
+	int err;
+	struct net_buf *buf;
+
+	err = bt_hci_register_vnd_evt_cb(on_vs_evt);
+	if (err) {
+		LOG_ERR("Failed to register HCI VS callback");
+		return;
+	}
+
+	hci_vs_cmd_qos_conn_event_report_enable_t *cmd_enable;
+
+	buf = bt_hci_cmd_create(HCI_VS_OPCODE_CMD_QOS_CONN_EVENT_REPORT_ENABLE,
+				sizeof(*cmd_enable));
+	if (!buf) {
+		LOG_ERR("Failed to enable HCI VS QoS");
+		return;
+	}
+
+	cmd_enable = net_buf_add(buf, sizeof(*cmd_enable));
+	cmd_enable->enable = 1;
+
+	err = bt_hci_cmd_send(
+		HCI_VS_OPCODE_CMD_QOS_CONN_EVENT_REPORT_ENABLE,
+		buf);
+	if (err) {
+		LOG_ERR("Failed to enable HCI VS QoS");
+		return;
+	}
+}
+
+static void handle_config_event(const struct config_event *event)
+{
+	switch (OPT_FIELD_GET(event->id)) {
+	case QOS_OPT_BLACKLIST:
+		if (event->dyndata.size != sizeof(struct params_blacklist)) {
+			LOG_WRN("Invalid size");
+		} else {
+			update_blacklist(event->dyndata.data);
+		}
+		break;
+	case QOS_OPT_CHMAP:
+		/* Avoid updating channel map directly to */
+		/* reduce complexity in interaction with */
+		/* chmap filter lib */
+		LOG_WRN("Not supported");
+		break;
+	case QOS_OPT_PARAM_BLE:
+		if (event->dyndata.size != sizeof(struct params_ble)) {
+			LOG_WRN("Invalid size");
+		} else {
+			update_parameters(event->dyndata.data, NULL);
+		}
+		break;
+	case QOS_OPT_PARAM_WIFI:
+		if (event->dyndata.size != sizeof(struct params_wifi)) {
+			LOG_WRN("Invalid size");
+		} else {
+			update_parameters(NULL, event->dyndata.data);
+		}
+		break;
+	default:
+		LOG_WRN("Unknown opt");
+		return;
+	}
+}
+
+static bool event_handler(const struct event_header *eh)
+{
+	if (IS_ENABLED(CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE) &&
+	    IS_ENABLED(CONFIG_DESKTOP_HID_MOUSE)) {
+		static s32_t hid_pkt_recv_count;
+		static u32_t cdc_notify_count;
+
+		/* Count number of HID packets received via BLE */
+		/* and number of HID packets sent via USB. */
+		/* Send stats printout via CDC every 100 packets */
+
+		if (is_hid_mouse_event(eh)) {
+			hid_pkt_recv_count++;
+			cdc_notify_count++;
+
+			if (cdc_notify_count == 100) {
+				hid_pkt_stats_print(
+					hid_pkt_recv_count);
+				cdc_notify_count = 0;
+			}
+
+			return false;
+		}
+	}
+
+	if (is_module_state_event(eh)) {
+		const struct module_state_event *event =
+			cast_module_state_event(eh);
+
+		if (check_state(
+			    event,
+			    MODULE_ID(main),
+			    MODULE_STATE_READY)) {
+			static bool initialized;
+			int err;
+
+			__ASSERT_NO_MSG(!initialized);
+
+			initialized = true;
+
+			chmap_filter_init();
+
+			chmap_inst =
+				(struct chmap_instance *) chmap_instance_buf;
+			err = chmap_filter_instance_init(
+				chmap_inst,
+				sizeof(chmap_instance_buf));
+			if (err) {
+				LOG_ERR("Failed to initialize filter");
+				__ASSERT_NO_MSG(err == 0);
```

Comment:
Maybe` module_set_state(MODULE_STATE_ERROR)` here ? This would indicate error also on Release configuration.

Reply:
Or it could be even better to properly disable the module, since application could run without it.
I think we should not `enable_qos_reporting` in such situation and we should not react on config channel events (so the user may notice that module is not working fine).